pkg/reconciler/workload/reconciler.go

```diff
@@ -0,0 +1,179 @@
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithPackager specifies how the Reconciler should package the workload.
+func WithPackager(p Packager) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.packager = p
+	}
+}
+
+// WithApplicator specifies how the Reconciler should apply the workload
+// package.
+func WithApplicator(a resource.Applicator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.applicator = a
+	}
+}
+
+// A Reconciler reconciles an OAM workload type by packaging it into a
+// KubernetesApplication.
+type Reconciler struct {
+	client      client.Client
+	newWorkload func() Workload
+	packager    Packager
+	applicator  resource.Applicator
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Kind is a kind of OAM workload.
+type Kind schema.GroupVersionKind
+
+// A Workload is a type of OAM workload.
+type Workload interface {
+	resource.Conditioned
+	metav1.Object
+	runtime.Object
+}
+
+// NewReconciler returns a Reconciler that reconciles an OAM workload type by
+// packaging it into a KubernetesApplication.
+func NewReconciler(m ctrl.Manager, workload Kind, o ...ReconcilerOption) *Reconciler {
+	nw := func() Workload {
+		return resource.MustCreateObject(schema.GroupVersionKind(workload), m.GetScheme()).(Workload)
+	}
+
+	r := &Reconciler{
+		client:      m.GetClient(),
+		newWorkload: nw,
+		packager:    PackageFn(genericPackager),
+		applicator:  resource.ApplyFn(resource.Apply),
+		log:         logging.NewNopLogger(),
+		record:      event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile an OAM workload type by packaging it into a KubernetesApplication.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	workload := r.newWorkload()
+	if err := r.client.Get(ctx, req.NamespacedName, workload); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetWorkload)
+	}
+
+	log = log.WithValues("uid", workload.GetUID(), "version", workload.GetResourceVersion())
+
+	app := &workloadv1alpha1.KubernetesApplication{}
+	if err := r.packager.Package(ctx, app, workload); err != nil {
+		log.Debug("Cannot package workload", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(workload, event.Warning(reasonCannotPackageWorkload, err))
+		workload.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errPackageWorkload)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, workload), errUpdateWorkloadStatus)
+	}
+
+	// A workload's package must have the same name and namespace, and must be
+	// controlled by the same owner.
+	app.SetOwnerReferences(workload.GetOwnerReferences())
```

Comment:
Should the app be owned by the owner of this workload, or by the workload itself?

Reply:
Likely by the workload itself since this controller is unaware of `ApplicationConfigurations` and should allow for direct creation of workloads as well :+1: